frontend/catalyst/pennylane_extensions.py

```diff
@@ -2633,22 +2635,38 @@ def callback_implementation(
     flat_args, in_tree = tree_flatten((args, kwargs))
     metadata = CallbackClosure(args, kwargs)
 
+    def to_shaped_array(ty):
+        if ty == inspect.Signature.empty:
+            return None
+        if isinstance(ty, ShapedArray):
+            return ty
+        return shaped_abstractify(ty)
+
+    results_aval = tree_map(to_shaped_array, result_shape_dtypes)
```

Comment:
Does tree_map map the function for all tree leaves?

Reply:
[Mm... not quite, but in our case, yes.](https://jax.readthedocs.io/en/latest/_autosummary/jax.tree_util.tree_map.html) The argument to `to_shaped_array` will always be a leaf, but it JAX has generalized the function and allowed `f` to take all the other leaves as well.

```
A new pytree with the same structure as tree but with the value at each leaf given by f(x, *xs) where x is the value at the corresponding leaf in tree and xs is the tuple of values at corresponding nodes in rest.
```